Add prepare-translations-for-npm action to core actions (#3)
BrightspaceUI/actions: prepare-translations-for-npm/README.md

> # Prepare Translations for NPM Action

Author: I don't know that I'm very good at writing `README`s and I have no idea if this is too much information or not enough. This started based on a copy/paste of the semantic-release action's README to try to align sections as much as possible.

This GitHub action prepares translation files in a project for NPM publishing by creating temporary `en` translations for each language term and each language
in your project that doesn't yet have a proper translation.

## Using the Action

> Typically this action is triggered from a workflow that runs on your `master` branch after each commit or pull request merge.

Reviewer: Is that true?  Wouldn't we typically run it only when we are about to release?

Author: I believe so. If this isn't followed by a publish to NPM, it's sort of useless. It makes changes to the translation files in the GHA workspace, but doesn't touch anything in the repo proper.

Reviewer 2: Maybe for clarity we say:
"Typically this action is triggered from a workflow that runs on your `master` branch after each commit or pull request merge that you are planning to publish to npm (for example, before running the `semantic-release` action <link to https://github.com/BrightspaceUI/actions/tree/master/semantic-release>)."

Author: This seems good to me.

> Typically this action is triggered from a workflow that runs on your `master` branch after each commit or pull request merge that you are planning to publish to npm (for example, before running the [`semantic-release` action](https://github.com/BrightspaceUI/actions/tree/master/semantic-release).

Reviewer 2: ```suggestion
Typically this action is triggered from a workflow that runs on your `master` branch after each commit or pull request merge that you are planning to publish to npm (for example, before running the [`semantic-release` action](https://github.com/BrightspaceUI/actions/tree/master/semantic-release)).

```

Author: Ah, thanks for the catch!

Here's a sample workflow:

```yml
name: Release
on:
  push:
    branches:
      - master
jobs:
  release:
    timeout-minutes: 2
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: Brightspace/third-party-actions@actions/checkout
        with:
          persist-credentials: false
      - name: Setup Node
        uses: Brightspace/third-party-actions@actions/setup-node
        # additional validation steps can be run here
      - name: Prepare Translations for NPM
        uses: BrightspaceUI/actions/prepare-translations-for-npm@master
        with:
          LANG_PATH: './lang'
        # NPM publishing step happens here
```

Options:
* `LANG_PATH`: The relative path to your language files from the root of your repository

Notes:

> * If you have additional release validation steps (e.g. build step, validation tests), run them before the "Prepare Translations for NPM" step.

Reviewer 3: I think this bullet could be excluded, since it's really more for the release actions.

Author: I mostly left it in because the "Prepare Translations" step does modify the working directory, which means tests technically won't be getting a "clean" checkout to work on. I don't know whether that matters in practice or whether it's common to write tests around translations. I'll happily yield to your experience here.

Reviewer 3: That's fair -- I'm good either way, more info is usually better!

# Prepare Translations for NPM Action

This GitHub action prepares translation files in a project for NPM publishing by creating temporary `en` translations for each language term and each language
in your project that doesn't yet have a proper translation.

## Using the Action

> Typically this action is triggered from a workflow that runs on your `master` branch after each commit or pull request merge that you are planning to publish to npm (for example, before running the [`semantic-release` action](https://github.com/BrightspaceUI/actions/tree/master/semantic-release)).

Author: ```suggestion
Typically this action is called from a release workflow before a step which publishes to NPM, such as the [semantic-release action](https://github.com/BrightspaceUI/actions/tree/master/semantic-release).
```

Author: Apparently I can't commit suggestions based on "outdated" diffs, so this is just your suggestion re-applied, @dlockhart.

# Prepare Translations for NPM Action

This GitHub action prepares translation files in a project for NPM publishing by creating temporary `en` translations for each language term and each language
in your project that doesn't yet have a proper translation.

## Using the Action

Typically this action is called from a release workflow before a step which publishes to NPM, such as the [semantic-release action](https://github.com/BrightspaceUI/actions/tree/master/semantic-release).

Here's a sample workflow:

```yml
name: Release
on:
  push:
    branches:
      - master
jobs:
  release:
    timeout-minutes: 2
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: Brightspace/third-party-actions@actions/checkout

>         with:

Reviewer 2: ```suggestion
```

Author: Is omitting `with:` valid when inputs are provided? I guess there's no explicit need to specify `persist-credentials` in this example workflow.

Reviewer 2: Oh maybe I did the wrong `with`

Author: Oh, did I leave it in for the Node piece? Whoops.

Reviewer 2: Yeah sorry, meant to put it on that one 🤦 

I don't mind having the `persist-credentials` here even though they don't need it for this, because they probably _would_ need it for the npm publish one (if they use our action)

Author: Yeah, I added it back and removed the extraneous `with`. I reconsidered and we may as well encourage use of our release actions.

      - name: Prepare Translations for NPM
        uses: BrightspaceUI/actions/prepare-translations-for-npm@master
        with:
          LANG_PATH: lang
        # NPM publishing step happens here
```

Options:
* `LANG_PATH` (default: `lang`): The relative path to your language files from the root of your repository

Notes:
* When setting up NodeJS, the version specified must be `>= 12`. This action uses features introduced in v12.
* If you have additional release validation steps (e.g. build step, validation tests), run them before the "Prepare Translations for NPM" step.
* You must run the checkout step before running the "Prepare Translations for NPM" step, otherwise the action will not be able to find your lang path.
* Don't run this action if you're not publishing to NPM. It doesn't commit to your repo and isn't relevant outside of publishing to NPM.

## Lang File Format

This action expects the following format for lang files:
* All lang files should be contained within the top level of the `LANG_PATH` directory. This action does not support language files within nested directories (e.g. `{LANG_PATH}/es/es.js`, `{LANG_PATH}/en/en.js` would not be supported).

> * Lang files must be JavaScript modules, formatted as outlined in the [localize mixin documentation](https://github.com/BrightspaceUI/core/blob/master/mixins/localize-mixin.md).

Author: I'm unsure if we have better documentation on this process, but if we do, I would be glad to link it here (or I can remove this reference if we think it's implied).

Author: I also added some extra info here that I thought was relevant (that empty default exports need to be present for supported langs, that lang files should go into the main list, etc.). I can remove or change them if there are objections.

Reviewer 2: So if we add a new language to the list, and they don't have a js file for it, does it fail the build with a helpful message?  Or we pass and warn?  Or we pass?

Author: We pass. Because this workflow is meant to come on release (post-build), it's challenging to get a message to the consumer. We could put it in the build log, but if the build passes, I'm not sure people will read it. I'm definitely open to having this work a different way, but I figured we would want to leave build failures out of this action in favour of (at some point) introducing a CI-level action that would actually fail the build if lang files were missing.

Reviewer 2: Yeah I think that's fine, just wasn't totally apparent from the wording and wanted to add info about failing if we did so.

Author: Ah, if it's not apparent from the wording, then I should make it so. I'll add some clarification there.

Reviewer 2: Awesome, that feels a lot clearer to me now
